File: binding/include/libqstyleditemdelegate.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qstyleditemdelegate.html
pub const qstyleditemdelegate = struct {

    /// New constructs a new QStyledItemDelegate object.
    pub fn New() QtC.QStyledItemDelegate {
        return qtc.QStyledItemDelegate_new();
    }


    /// New2 constructs a new QStyledItemDelegate object.
    pub fn New2(parent: ?*anyopaque) QtC.QStyledItemDelegate {
        return qtc.QStyledItemDelegate_new2(@ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QStyledItemDelegate_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qstyleditemdelegate.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Paint(self: ?*anyopaque, painter: ?*anyopaque, option: ?*anyopaque, index: ?*anyopaque) void {
        qtc.QStyledItemDelegate_Paint(@ptrCast(self), @ptrCast(painter), @ptrCast(option), @ptrCast(index));
    }

    /// Allows for overriding the related default method
    pub fn OnPaint(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque, ?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QStyledItemDelegate_OnPaint(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBasePaint(self: ?*anyopaque, painter: ?*anyopaque, option: ?*anyopaque, index: ?*anyopaque) void {
        qtc.QStyledItemDelegate_QBasePaint(@ptrCast(self), @ptrCast(painter), @ptrCast(option), @ptrCast(index));
    }

    pub fn SizeHint(self: ?*anyopaque, option: ?*anyopaque, index: ?*anyopaque) QtC.QSize {
        return qtc.QStyledItemDelegate_SizeHint(@ptrCast(self), @ptrCast(option), @ptrCast(index));
    }

    /// Allows for overriding the related default method
    pub fn OnSizeHint(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque, ?*anyopaque) callconv(.c) QtC.QSize) void {
        qtc.QStyledItemDelegate_OnSizeHint(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseSizeHint(self: ?*anyopaque, option: ?*anyopaque, index: ?*anyopaque) QtC.QSize {
        return qtc.QStyledItemDelegate_QBaseSizeHint(@ptrCast(self), @ptrCast(option), @ptrCast(index));
    }

    pub fn CreateEditor(self: ?*anyopaque, parent: ?*anyopaque, option: ?*anyopaque, index: ?*anyopaque) QtC.QWidget {
        return qtc.QStyledItemDelegate_CreateEditor(@ptrCast(self), @ptrCast(parent), @ptrCast(option), @ptrCast(index));
    }

    /// Allows for overriding the related default method
    pub fn OnCreateEditor(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque, ?*anyopaque, ?*anyopaque) callconv(.c) QtC.QWidget) void {
        qtc.QStyledItemDelegate_OnCreateEditor(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseCreateEditor(self: ?*anyopaque, parent: ?*anyopaque, option: ?*anyopaque, index: ?*anyopaque) QtC.QWidget {
        return qtc.QStyledItemDelegate_QBaseCreateEditor(@ptrCast(self), @ptrCast(parent), @ptrCast(option), @ptrCast(index));
    }

    pub fn SetEditorData(self: ?*anyopaque, editor: ?*anyopaque, index: ?*anyopaque) void {
        qtc.QStyledItemDelegate_SetEditorData(@ptrCast(self), @ptrCast(editor), @ptrCast(index));
    }

    /// Allows for overriding the related default method
    pub fn OnSetEditorData(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QStyledItemDelegate_OnSetEditorData(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseSetEditorData(self: ?*anyopaque, editor: ?*anyopaque, index: ?*anyopaque) void {
        qtc.QStyledItemDelegate_QBaseSetEditorData(@ptrCast(self), @ptrCast(editor), @ptrCast(index));
    }

    pub fn SetModelData(self: ?*anyopaque, editor: ?*anyopaque, model: ?*anyopaque, index: ?*anyopaque) void {
        qtc.QStyledItemDelegate_SetModelData(@ptrCast(self), @ptrCast(editor), @ptrCast(model), @ptrCast(index));
    }

    /// Allows for overriding the related default method
    pub fn OnSetModelData(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque, ?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QStyledItemDelegate_OnSetModelData(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseSetModelData(self: ?*anyopaque, editor: ?*anyopaque, model: ?*anyopaque, index: ?*anyopaque) void {
        qtc.QStyledItemDelegate_QBaseSetModelData(@ptrCast(self), @ptrCast(editor), @ptrCast(model), @ptrCast(index));
    }

    pub fn UpdateEditorGeometry(self: ?*anyopaque, editor: ?*anyopaque, option: ?*anyopaque, index: ?*anyopaque) void {
        qtc.QStyledItemDelegate_UpdateEditorGeometry(@ptrCast(self), @ptrCast(editor), @ptrCast(option), @ptrCast(index));
    }

    /// Allows for overriding the related default method
    pub fn OnUpdateEditorGeometry(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque, ?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QStyledItemDelegate_OnUpdateEditorGeometry(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseUpdateEditorGeometry(self: ?*anyopaque, editor: ?*anyopaque, option: ?*anyopaque, index: ?*anyopaque) void {
        qtc.QStyledItemDelegate_QBaseUpdateEditorGeometry(@ptrCast(self), @ptrCast(editor), @ptrCast(option), @ptrCast(index));
    }

    pub fn ItemEditorFactory(self: ?*anyopaque, ) QtC.QItemEditorFactory {
        return qtc.QStyledItemDelegate_ItemEditorFactory(@ptrCast(self));
    }

    pub fn SetItemEditorFactory(self: ?*anyopaque, factory: ?*anyopaque) void {
        qtc.QStyledItemDelegate_SetItemEditorFactory(@ptrCast(self), @ptrCast(factory));
    }

    pub fn DisplayText(self: ?*anyopaque, value: ?*anyopaque, locale: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QStyledItemDelegate_DisplayText(@ptrCast(self), @ptrCast(value), @ptrCast(locale));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qstyleditemdelegate.DisplayText: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Allows for overriding the related default method
    pub fn OnDisplayText(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque, ?*anyopaque) callconv(.c) [*:0]const u8) void {
        qtc.QStyledItemDelegate_OnDisplayText(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseDisplayText(self: ?*anyopaque, value: ?*anyopaque, locale: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QStyledItemDelegate_QBaseDisplayText(@ptrCast(self), @ptrCast(value), @ptrCast(locale));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qstyleditemdelegate.DisplayText: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn InitStyleOption(self: ?*anyopaque, option: ?*anyopaque, index: ?*anyopaque) void {
        qtc.QStyledItemDelegate_InitStyleOption(@ptrCast(self), @ptrCast(option), @ptrCast(index));
    }

    /// Allows for overriding the related default method
    pub fn OnInitStyleOption(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QStyledItemDelegate_OnInitStyleOption(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseInitStyleOption(self: ?*anyopaque, option: ?*anyopaque, index: ?*anyopaque) void {
        qtc.QStyledItemDelegate_QBaseInitStyleOption(@ptrCast(self), @ptrCast(option), @ptrCast(index));
    }

    pub fn EventFilter(self: ?*anyopaque, object: ?*anyopaque, event: ?*anyopaque) bool {
        return qtc.QStyledItemDelegate_EventFilter(@ptrCast(self), @ptrCast(object), @ptrCast(event));
    }

    /// Allows for overriding the related default method
    pub fn OnEventFilter(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque, ?*anyopaque) callconv(.c) bool) void {
        qtc.QStyledItemDelegate_OnEventFilter(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseEventFilter(self: ?*anyopaque, object: ?*anyopaque, event: ?*anyopaque) bool {
        return qtc.QStyledItemDelegate_QBaseEventFilter(@ptrCast(self), @ptrCast(object), @ptrCast(event));
    }

    pub fn EditorEvent(self: ?*anyopaque, event: ?*anyopaque, model: ?*anyopaque, option: ?*anyopaque, index: ?*anyopaque) bool {
        return qtc.QStyledItemDelegate_EditorEvent(@ptrCast(self), @ptrCast(event), @ptrCast(model), @ptrCast(option), @ptrCast(index));
    }

    /// Allows for overriding the related default method
    pub fn OnEditorEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque, ?*anyopaque, ?*anyopaque, ?*anyopaque) callconv(.c) bool) void {
        qtc.QStyledItemDelegate_OnEditorEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseEditorEvent(self: ?*anyopaque, event: ?*anyopaque, model: ?*anyopaque, option: ?*anyopaque, index: ?*anyopaque) bool {
        return qtc.QStyledItemDelegate_QBaseEditorEvent(@ptrCast(self), @ptrCast(event), @ptrCast(model), @ptrCast(option), @ptrCast(index));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QStyledItemDelegate_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qstyleditemdelegate.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QStyledItemDelegate_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qstyleditemdelegate.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QStyledItemDelegate_Delete(@ptrCast(self));
    }
};

File: binding/src/libqstyleditemdelegate.cpp
#include <QAbstractItemDelegate>
#include <QAbstractItemModel>
#include <QEvent>
#include <QItemEditorFactory>
#include <QLocale>
#include <QModelIndex>
#include <QObject>
#include <QPainter>
#include <QSize>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QStyleOptionViewItem>
#include <QStyledItemDelegate>
#include <QVariant>
#include <QWidget>
#include <qstyleditemdelegate.h>
#include "libqstyleditemdelegate.h"
#include "libqstyleditemdelegate.hxx"

QStyledItemDelegate* QStyledItemDelegate_new() {
	 return new VirtualQStyledItemDelegate();
}

QStyledItemDelegate* QStyledItemDelegate_new2(QObject* parent) {
	 return new VirtualQStyledItemDelegate(parent);
}

libqt_string QStyledItemDelegate_Tr(const char* s) {
	QString _ret = QStyledItemDelegate::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QStyledItemDelegate_Paint(const QStyledItemDelegate* self, QPainter* painter, const QStyleOptionViewItem* option, const QModelIndex* index) {
	self->paint(painter, *option, *index);
}

QSize* QStyledItemDelegate_SizeHint(const QStyledItemDelegate* self, const QStyleOptionViewItem* option, const QModelIndex* index) {
	return new QSize(self->sizeHint(*option, *index));
}

QWidget* QStyledItemDelegate_CreateEditor(const QStyledItemDelegate* self, QWidget* parent, const QStyleOptionViewItem* option, const QModelIndex* index) {
	return self->createEditor(parent, *option, *index);
}

void QStyledItemDelegate_SetEditorData(const QStyledItemDelegate* self, QWidget* editor, const QModelIndex* index) {
	self->setEditorData(editor, *index);
}

void QStyledItemDelegate_SetModelData(const QStyledItemDelegate* self, QWidget* editor, QAbstractItemModel* model, const QModelIndex* index) {
	self->setModelData(editor, model, *index);
}

void QStyledItemDelegate_UpdateEditorGeometry(const QStyledItemDelegate* self, QWidget* editor, const QStyleOptionViewItem* option, const QModelIndex* index) {
	self->updateEditorGeometry(editor, *option, *index);
}

QItemEditorFactory* QStyledItemDelegate_ItemEditorFactory(const QStyledItemDelegate* self) {
	return self->itemEditorFactory();
}

void QStyledItemDelegate_SetItemEditorFactory(QStyledItemDelegate* self, QItemEditorFactory* factory) {
	self->setItemEditorFactory(factory);
}

libqt_string QStyledItemDelegate_DisplayText(const QStyledItemDelegate* self, const QVariant* value, const QLocale* locale) {
	QString _ret = self->displayText(*value, *locale);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QStyledItemDelegate_Tr2(const char* s, const char* c) {
	QString _ret = QStyledItemDelegate::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QStyledItemDelegate_Tr3(const char* s, const char* c, int n) {
	QString _ret = QStyledItemDelegate::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

// Base class handler implementation
void QStyledItemDelegate_QBasePaint(const QStyledItemDelegate* self, QPainter* painter, const QStyleOptionViewItem* option, const QModelIndex* index) {
	auto* vqstyleditemdelegate = dynamic_cast<const VirtualQStyledItemDelegate*>(self);
	if (vqstyleditemdelegate && vqstyleditemdelegate->isVirtualQStyledItemDelegate) {
vqstyleditemdelegate->setQStyledItemDelegate_Paint_IsBase(true);
	vqstyleditemdelegate->paint(painter, *option, *index);
}
}

// Auxiliary method to allow providing re-implementation
void QStyledItemDelegate_OnPaint(const QStyledItemDelegate* self, intptr_t slot) {
	auto* vqstyleditemdelegate = dynamic_cast<const VirtualQStyledItemDelegate*>(self);
	if (vqstyleditemdelegate && vqstyleditemdelegate->isVirtualQStyledItemDelegate) {
vqstyleditemdelegate->setQStyledItemDelegate_Paint_Callback(reinterpret_cast<VirtualQStyledItemDelegate::QStyledItemDelegate_Paint_Callback>(slot));
}
}

// Base class handler implementation
QSize* QStyledItemDelegate_QBaseSizeHint(const QStyledItemDelegate* self, const QStyleOptionViewItem* option, const QModelIndex* index) {
	auto* vqstyleditemdelegate = dynamic_cast<const VirtualQStyledItemDelegate*>(self);
	if (vqstyleditemdelegate && vqstyleditemdelegate->isVirtualQStyledItemDelegate) {
vqstyleditemdelegate->setQStyledItemDelegate_SizeHint_IsBase(true);
	return new QSize(vqstyleditemdelegate->sizeHint(*option, *index));
}
}

// Auxiliary method to allow providing re-implementation
void QStyledItemDelegate_OnSizeHint(const QStyledItemDelegate* self, intptr_t slot) {
	auto* vqstyleditemdelegate = dynamic_cast<const VirtualQStyledItemDelegate*>(self);
	if (vqstyleditemdelegate && vqstyleditemdelegate->isVirtualQStyledItemDelegate) {
vqstyleditemdelegate->setQStyledItemDelegate_SizeHint_Callback(reinterpret_cast<VirtualQStyledItemDelegate::QStyledItemDelegate_SizeHint_Callback>(slot));
}
}

// Base class handler implementation
QWidget* QStyledItemDelegate_QBaseCreateEditor(const QStyledItemDelegate* self, QWidget* parent, const QStyleOptionViewItem* option, const QModelIndex* index) {
	auto* vqstyleditemdelegate = dynamic_cast<const VirtualQStyledItemDelegate*>(self);
	if (vqstyleditemdelegate && vqstyleditemdelegate->isVirtualQStyledItemDelegate) {
vqstyleditemdelegate->setQStyledItemDelegate_CreateEditor_IsBase(true);
	return vqstyleditemdelegate->createEditor(parent, *option, *index);
}
}

// Auxiliary method to allow providing re-implementation
void QStyledItemDelegate_OnCreateEditor(const QStyledItemDelegate* self, intptr_t slot) {
	auto* vqstyleditemdelegate = dynamic_cast<const VirtualQStyledItemDelegate*>(self);
	if (vqstyleditemdelegate && vqstyleditemdelegate->isVirtualQStyledItemDelegate) {
vqstyleditemdelegate->setQStyledItemDelegate_CreateEditor_Callback(reinterpret_cast<VirtualQStyledItemDelegate::QStyledItemDelegate_CreateEditor_Callback>(slot));
}
}

// Base class handler implementation
void QStyledItemDelegate_QBaseSetEditorData(const QStyledItemDelegate* self, QWidget* editor, const QModelIndex* index) {
	auto* vqstyleditemdelegate = dynamic_cast<const VirtualQStyledItemDelegate*>(self);
	if (vqstyleditemdelegate && vqstyleditemdelegate->isVirtualQStyledItemDelegate) {
vqstyleditemdelegate->setQStyledItemDelegate_SetEditorData_IsBase(true);
	vqstyleditemdelegate->setEditorData(editor, *index);
}
}

// Auxiliary method to allow providing re-implementation
void QStyledItemDelegate_OnSetEditorData(const QStyledItemDelegate* self, intptr_t slot) {
	auto* vqstyleditemdelegate = dynamic_cast<const VirtualQStyledItemDelegate*>(self);
	if (vqstyleditemdelegate && vqstyleditemdelegate->isVirtualQStyledItemDelegate) {
vqstyleditemdelegate->setQStyledItemDelegate_SetEditorData_Callback(reinterpret_cast<VirtualQStyledItemDelegate::QStyledItemDelegate_SetEditorData_Callback>(slot));
}
}

// Base class handler implementation
void QStyledItemDelegate_QBaseSetModelData(const QStyledItemDelegate* self, QWidget* editor, QAbstractItemModel* model, const QModelIndex* index) {
	auto* vqstyleditemdelegate = dynamic_cast<const VirtualQStyledItemDelegate*>(self);
	if (vqstyleditemdelegate && vqstyleditemdelegate->isVirtualQStyledItemDelegate) {
vqstyleditemdelegate->setQStyledItemDelegate_SetModelData_IsBase(true);
	vqstyleditemdelegate->setModelData(editor, model, *index);
}
}

// Auxiliary method to allow providing re-implementation
void QStyledItemDelegate_OnSetModelData(const QStyledItemDelegate* self, intptr_t slot) {
	auto* vqstyleditemdelegate = dynamic_cast<const VirtualQStyledItemDelegate*>(self);
	if (vqstyleditemdelegate && vqstyleditemdelegate->isVirtualQStyledItemDelegate) {
vqstyleditemdelegate->setQStyledItemDelegate_SetModelData_Callback(reinterpret_cast<VirtualQStyledItemDelegate::QStyledItemDelegate_SetModelData_Callback>(slot));
}
}

// Base class handler implementation
void QStyledItemDelegate_QBaseUpdateEditorGeometry(const QStyledItemDelegate* self, QWidget* editor, const QStyleOptionViewItem* option, const QModelIndex* index) {
	auto* vqstyleditemdelegate = dynamic_cast<const VirtualQStyledItemDelegate*>(self);
	if (vqstyleditemdelegate && vqstyleditemdelegate->isVirtualQStyledItemDelegate) {
vqstyleditemdelegate->setQStyledItemDelegate_UpdateEditorGeometry_IsBase(true);
	vqstyleditemdelegate->updateEditorGeometry(editor, *option, *index);
}
}

// Auxiliary method to allow providing re-implementation
void QStyledItemDelegate_OnUpdateEditorGeometry(const QStyledItemDelegate* self, intptr_t slot) {
	auto* vqstyleditemdelegate = dynamic_cast<const VirtualQStyledItemDelegate*>(self);
	if (vqstyleditemdelegate && vqstyleditemdelegate->isVirtualQStyledItemDelegate) {
vqstyleditemdelegate->setQStyledItemDelegate_UpdateEditorGeometry_Callback(reinterpret_cast<VirtualQStyledItemDelegate::QStyledItemDelegate_UpdateEditorGeometry_Callback>(slot));
}
}

// Base class handler implementation
libqt_string QStyledItemDelegate_QBaseDisplayText(const QStyledItemDelegate* self, const QVariant* value, const QLocale* locale) {
	auto* vqstyleditemdelegate = dynamic_cast<const VirtualQStyledItemDelegate*>(self);
	if (vqstyleditemdelegate && vqstyleditemdelegate->isVirtualQStyledItemDelegate) {
vqstyleditemdelegate->setQStyledItemDelegate_DisplayText_IsBase(true);
	QString _ret = vqstyleditemdelegate->displayText(*value, *locale);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}
}

// Auxiliary method to allow providing re-implementation
void QStyledItemDelegate_OnDisplayText(const QStyledItemDelegate* self, intptr_t slot) {
	auto* vqstyleditemdelegate = dynamic_cast<const VirtualQStyledItemDelegate*>(self);
	if (vqstyleditemdelegate && vqstyleditemdelegate->isVirtualQStyledItemDelegate) {
vqstyleditemdelegate->setQStyledItemDelegate_DisplayText_Callback(reinterpret_cast<VirtualQStyledItemDelegate::QStyledItemDelegate_DisplayText_Callback>(slot));
}
}

// Derived class handler implementation
void QStyledItemDelegate_InitStyleOption(const QStyledItemDelegate* self, QStyleOptionViewItem* option, const QModelIndex* index) {
	auto* vqstyleditemdelegate = dynamic_cast<const VirtualQStyledItemDelegate*>(self);
	if (vqstyleditemdelegate && vqstyleditemdelegate->isVirtualQStyledItemDelegate) {
	vqstyleditemdelegate->initStyleOption(option, *index);
	} else {
	self->QStyledItemDelegate::initStyleOption(option, *index);
}
}

// Base class handler implementation
void QStyledItemDelegate_QBaseInitStyleOption(const QStyledItemDelegate* self, QStyleOptionViewItem* option, const QModelIndex* index) {
	auto* vqstyleditemdelegate = dynamic_cast<const VirtualQStyledItemDelegate*>(self);
	if (vqstyleditemdelegate && vqstyleditemdelegate->isVirtualQStyledItemDelegate) {
vqstyleditemdelegate->setQStyledItemDelegate_InitStyleOption_IsBase(true);
	vqstyleditemdelegate->initStyleOption(option, *index);
}
}

// Auxiliary method to allow providing re-implementation
void QStyledItemDelegate_OnInitStyleOption(const QStyledItemDelegate* self, intptr_t slot) {
	auto* vqstyleditemdelegate = dynamic_cast<const VirtualQStyledItemDelegate*>(self);
	if (vqstyleditemdelegate && vqstyleditemdelegate->isVirtualQStyledItemDelegate) {
vqstyleditemdelegate->setQStyledItemDelegate_InitStyleOption_Callback(reinterpret_cast<VirtualQStyledItemDelegate::QStyledItemDelegate_InitStyleOption_Callback>(slot));
}
}

// Derived class handler implementation
bool QStyledItemDelegate_EventFilter(QStyledItemDelegate* self, QObject* object, QEvent* event) {
	auto* vqstyleditemdelegate = dynamic_cast<VirtualQStyledItemDelegate*>(self);
	if (vqstyleditemdelegate && vqstyleditemdelegate->isVirtualQStyledItemDelegate) {
	return vqstyleditemdelegate->eventFilter(object, event);
	} else {
	return self->QStyledItemDelegate::eventFilter(object, event);
}
}

// Base class handler implementation
bool QStyledItemDelegate_QBaseEventFilter(QStyledItemDelegate* self, QObject* object, QEvent* event) {
	auto* vqstyleditemdelegate = dynamic_cast<VirtualQStyledItemDelegate*>(self);
	if (vqstyleditemdelegate && vqstyleditemdelegate->isVirtualQStyledItemDelegate) {
vqstyleditemdelegate->setQStyledItemDelegate_EventFilter_IsBase(true);
	return vqstyleditemdelegate->eventFilter(object, event);
}
}

// Auxiliary method to allow providing re-implementation
void QStyledItemDelegate_OnEventFilter(QStyledItemDelegate* self, intptr_t slot) {
	auto* vqstyleditemdelegate = dynamic_cast<VirtualQStyledItemDelegate*>(self);
	if (vqstyleditemdelegate && vqstyleditemdelegate->isVirtualQStyledItemDelegate) {
vqstyleditemdelegate->setQStyledItemDelegate_EventFilter_Callback(reinterpret_cast<VirtualQStyledItemDelegate::QStyledItemDelegate_EventFilter_Callback>(slot));
}
}

// Derived class handler implementation
bool QStyledItemDelegate_EditorEvent(QStyledItemDelegate* self, QEvent* event, QAbstractItemModel* model, const QStyleOptionViewItem* option, const QModelIndex* index) {
	auto* vqstyleditemdelegate = dynamic_cast<VirtualQStyledItemDelegate*>(self);
	if (vqstyleditemdelegate && vqstyleditemdelegate->isVirtualQStyledItemDelegate) {
	return vqstyleditemdelegate->editorEvent(event, model, *option, *index);
	} else {
	return self->QStyledItemDelegate::editorEvent(event, model, *option, *index);
}
}

// Base class handler implementation
bool QStyledItemDelegate_QBaseEditorEvent(QStyledItemDelegate* self, QEvent* event, QAbstractItemModel* model, const QStyleOptionViewItem* option, const QModelIndex* index) {
	auto* vqstyleditemdelegate = dynamic_cast<VirtualQStyledItemDelegate*>(self);
	if (vqstyleditemdelegate && vqstyleditemdelegate->isVirtualQStyledItemDelegate) {
vqstyleditemdelegate->setQStyledItemDelegate_EditorEvent_IsBase(true);
	return vqstyleditemdelegate->editorEvent(event, model, *option, *index);
}
}

// Auxiliary method to allow providing re-implementation
void QStyledItemDelegate_OnEditorEvent(QStyledItemDelegate* self, intptr_t slot) {
	auto* vqstyleditemdelegate = dynamic_cast<VirtualQStyledItemDelegate*>(self);
	if (vqstyleditemdelegate && vqstyleditemdelegate->isVirtualQStyledItemDelegate) {
vqstyleditemdelegate->setQStyledItemDelegate_EditorEvent_Callback(reinterpret_cast<VirtualQStyledItemDelegate::QStyledItemDelegate_EditorEvent_Callback>(slot));
}
}

void QStyledItemDelegate_Delete(QStyledItemDelegate* self) {
    delete self;
}

